In [1]:
import pandas as pd

# Load the data.
url1="https://bd29ee0e-54ab-4daa-9671-d153865d1620.usrfiles.com/ugd/bd29ee_c5bd3eaccf14437180eb3f730f718de9.csv"
df1=pd.read_csv(url1,encoding = "latin1")

url2="https://bd29ee0e-54ab-4daa-9671-d153865d1620.usrfiles.com/ugd/bd29ee_229df9e0e5b24090b8985434abf5a83d.csv"
df2=pd.read_csv(url2,encoding = 'latin1')

In [2]:
df1.head()

Date  ...  Series_Complete_FedLTC_Unknown
0  6/27/2021  ...                            4863
1  6/27/2021  ...                           36234
2  6/27/2021  ...                           31667
3  6/27/2021  ...                               0
4  6/27/2021  ...                           14681

[5 rows x 70 columns]

In [3]:
# Extract the cumulative statistics for vaccination progress.
df1=df1[::-1]
df3=df1[df1["Date"]=="6/26/2021"]
df3=df3[["Date","Location","State","Administered_Dose1_Pop_Pct","Series_Complete_Pop_Pct"]].sort_values(by=["State"]).reset_index(drop=True)
df3

Date Location  ... Administered_Dose1_Pop_Pct  Series_Complete_Pop_Pct
0   6/26/2021       AL  ...                       39.4                     32.2
1   6/26/2021       AK  ...                       48.4                     42.4
2   6/26/2021       AS  ...                       45.8                     37.7
3   6/26/2021       AZ  ...                       49.2                     39.6
4   6/26/2021       AR  ...                       41.6                     33.8
5   6/26/2021       CA  ...                       60.8                     49.0
6   6/26/2021       CO  ...                       57.6                     51.1
7   6/26/2021       CT  ...                       66.6                     59.7
8   6/26/2021       DE  ...                       57.7                     48.6
9   6/26/2021       DC  ...                       60.7                     51.5
10  6/26/2021       FL  ...                       53.1                     44.9
11  6/26/2021       GA  ...                       42.5                     35.4
12  6/26/2021       GU  ...                       60.9                     53.4
13  6/26/2021       HI  ...                       69.4                     51.5
14  6/26/2021       ID  ...                       39.3                     35.6
15  6/26/2021       IL  ...                       58.8                     45.0
16  6/26/2021       IN  ...                       44.3                     39.5
17  6/26/2021       IA  ...                       51.2                     47.5
18  6/26/2021       KS  ...                       48.9                     41.4
19  6/26/2021       KY  ...                       49.2                     42.8
20  6/26/2021       LA  ...                       37.8                     34.2
21  6/26/2021       ME  ...                       66.1                     60.7
22  6/26/2021       MD  ...                       61.0                     54.6
23  6/26/2021       MA  ...                       70.0                     60.7
24  6/26/2021       MI  ...                       51.2                     46.6
25  6/26/2021       MN  ...                       56.7                     50.8
26  6/26/2021       MS  ...                       35.9                     29.3
27  6/26/2021       MO  ...                       44.5                     38.5
28  6/26/2021       MT  ...                       47.5                     42.0
29  6/26/2021       NE  ...                       51.4                     47.1
30  6/26/2021       NV  ...                       48.9                     41.1
31  6/26/2021       NH  ...                       61.8                     55.2
32  6/26/2021       NJ  ...                       62.4                     54.2
33  6/26/2021       NM  ...                       61.6                     53.4
34  6/26/2021       NY  ...                       59.6                     53.1
35  6/26/2021       NC  ...                       45.0                     39.0
36  6/26/2021       ND  ...                       43.7                     38.6
37  6/26/2021       MP  ...                       51.2                     45.8
38  6/26/2021       OH  ...                       48.0                     44.2
39  6/26/2021       OK  ...                       44.6                     37.9
40  6/26/2021       OR  ...                       58.2                     52.3
41  6/26/2021       PA  ...                       62.4                     49.2
42  6/26/2021       PR  ...                       57.5                     45.1
43  6/26/2021       RI  ...                       64.2                     58.1
44  6/26/2021       SC  ...                       43.8                     37.8
45  6/26/2021       SD  ...                       50.3                     45.0
46  6/26/2021       TN  ...                       41.3                     34.9
47  6/26/2021       TX  ...                       47.8                     40.5
48  6/26/2021       UT  ...                       48.1                     36.7
49  6/26/

In [4]:
df2.tail()

date          state code  fips   cases  deaths
26464  6/26/2021       Virginia   VA    51  679680   11392
26465  6/26/2021     Washington   WA    53  452327    5948
26466  6/26/2021  West Virginia   WV    54  163804    2876
26467  6/26/2021      Wisconsin   WI    55  677464    8099
26468  6/26/2021        Wyoming   WY    56   61951     740

##### The goal is to analyze whether and how the availability of vaccines has impacted COVID-19 deaths. Hence, when we look at death rate, we will look at the data after vaccine has been made available to everyone, not the death rate since the onset of the pandemic. To extract only the data since 5/1/2021, we are going to subtract the statistics on 4/30/2021 from the cumulative data for the latest date. In other words, we are going to use case number and death toll after 5/1/2021.

In [5]:
sub_df2=df2[df2["date"]=="4/30/2021"].sort_values(by=["state"]).reset_index(drop=True)

In [6]:
df4=df2[df2["date"]=="6/26/2021"].sort_values(by=["state"]).reset_index(drop=True)
df4["cases"]=df4["cases"]-sub_df2["cases"]
df4["deaths"]=df4["deaths"]-sub_df2["deaths"]
df4

date                     state code  fips  cases  deaths
0   6/26/2021                   Alabama   AL     1  22012     440
1   6/26/2021                    Alaska   AK     2   2887      27
2   6/26/2021                   Arizona   AZ     4  30214     590
3   6/26/2021                  Arkansas   AR     5  11529     154
4   6/26/2021                California   CA     6  72954    1799
5   6/26/2021                  Colorado   CO     8  45888     487
6   6/26/2021               Connecticut   CT     9   9887     178
7   6/26/2021                  Delaware   DE    10   5252      69
8   6/26/2021      District of Columbia   DC    11   1608      36
9   6/26/2021                   Florida   FL    12  88419    2612
10  6/26/2021                   Georgia   GA    13  32814    1138
11  6/26/2021                      Guam   GU    66    364       3
12  6/26/2021                    Hawaii   HI    15   3484      33
13  6/26/2021                     Idaho   ID    16   7042      99
14  6/26/2021                  Illinois   IL    17  55767    1346
15  6/26/2021                   Indiana   IN    18  32305     490
16  6/26/2021                      Iowa   IA    19   8688     181
17  6/26/2021                    Kansas   KS    20   7620     157
18  6/26/2021                  Kentucky   KY    21  17759     632
19  6/26/2021                 Louisiana   LA    22  20940     341
20  6/26/2021                     Maine   ME    23   7750      74
21  6/26/2021                  Maryland   MD    24  14705     990
22  6/26/2021             Massachusetts   MA    25  20765     379
23  6/26/2021                  Michigan   MI    26  63116    2202
24  6/26/2021                 Minnesota   MN    27  29228     434
25  6/26/2021               Mississippi   MS    28   8937     196
26  6/26/2021                  Missouri   MO    29  30514     645
27  6/26/2021                   Montana   MT    30   4705      90
28  6/26/2021                  Nebraska   NE    31   4400     153
29  6/26/2021                    Nevada   NV    32  16176     203
30  6/26/2021             New Hampshire   NH    33   4517      68
31  6/26/2021                New Jersey   NJ    34  25458     880
32  6/26/2021                New Mexico   NM    35   7557     268
33  6/26/2021                  New York   NY    36  65611    1338
34  6/26/2021            North Carolina   NC    37  42683     756
35  6/26/2021              North Dakota   ND    38   3184      36
36  6/26/2021  Northern Mariana Islands   MP    69     19       0
37  6/26/2021                      Ohio   OH    39  38234     997
38  6/26/2021                  Oklahoma   OK    40   8605     596
39  6/26/2021                    Oregon   OR    41  23233     283
40  6/26/2021              Pennsylvania   PA    42  59790    1416
41  6/26/2021               Puerto Rico   PR    72   8472     243
42  6/26/2021              Rhode Island   RI    44   4328      57
43  6/26/2021            South Carolina   SC    45  17485     322
44  6/26/2021              South Dakota   SD    46   1832      63
45  6/26/2021                 Tennessee   TN    47  20933     349
46  6/26/2021                     Texas   TX    48  98883    2046
47  6/26/2021                      Utah   UT    49  16351     153
48  6/26/2021                   Vermont   VT    50   1440      10
49  6/26/2021            Virgin Islands   VI    78    710       3
50  6/26/2021                  Virginia   VA    51  20090     622
51  6/26/2021                Washington   WA    53  47355     401
52  6/26/2021             West Virginia   WV    54  10676     197
53  6/26/2021                 Wisconsin   WI    55  16802     549
54  6/26/2021                   Wyoming   WY    56   3809      33

##### There is some data cleaning that needs to be done. The extracted data for vaccination progress has 56 rows whereas the dataframe for case number and death toll since 5/1/2021 has only 55 rows. That means there is an extra row in the vaccination progress dataframe. We are going to rename the "Location" column to "code" on the vaccination progress dataf, then merge the two dataframes together using "inner join" and key "code".

In [8]:
df3=df3.rename(columns={"Location":"code"})
df5=pd.merge(df3, df4, on='code', how='inner')
df5

Date code                 State  ...  fips  cases deaths
0   6/26/2021   AL               ALABAMA  ...     1  22012    440
1   6/26/2021   AK                ALASKA  ...     2   2887     27
2   6/26/2021   AZ               ARIZONA  ...     4  30214    590
3   6/26/2021   AR              ARKANSAS  ...     5  11529    154
4   6/26/2021   CA            CALIFORNIA  ...     6  72954   1799
5   6/26/2021   CO              COLORADO  ...     8  45888    487
6   6/26/2021   CT           CONNECTICUT  ...     9   9887    178
7   6/26/2021   DE              DELAWARE  ...    10   5252     69
8   6/26/2021   DC  DISTRICT OF COLUMBIA  ...    11   1608     36
9   6/26/2021   FL               FLORIDA  ...    12  88419   2612
10  6/26/2021   GA               GEORGIA  ...    13  32814   1138
11  6/26/2021   GU                  GUAM  ...    66    364      3
12  6/26/2021   HI                HAWAII  ...    15   3484     33
13  6/26/2021   ID                 IDAHO  ...    16   7042     99
14  6/26/2021   IL              ILLINOIS  ...    17  55767   1346
15  6/26/2021   IN               INDIANA  ...    18  32305    490
16  6/26/2021   IA                  IOWA  ...    19   8688    181
17  6/26/2021   KS                KANSAS  ...    20   7620    157
18  6/26/2021   KY              KENTUCKY  ...    21  17759    632
19  6/26/2021   LA             LOUISIANA  ...    22  20940    341
20  6/26/2021   ME                 MAINE  ...    23   7750     74
21  6/26/2021   MD              MARYLAND  ...    24  14705    990
22  6/26/2021   MA         MASSACHUSETTS  ...    25  20765    379
23  6/26/2021   MI              MICHIGAN  ...    26  63116   2202
24  6/26/2021   MN             MINNESOTA  ...    27  29228    434
25  6/26/2021   MS           MISSISSIPPI  ...    28   8937    196
26  6/26/2021   MO              MISSOURI  ...    29  30514    645
27  6/26/2021   MT               MONTANA  ...    30   4705     90
28  6/26/2021   NE              NEBRASKA  ...    31   4400    153
29  6/26/2021   NV                NEVADA  ...    32  16176    203
30  6/26/2021   NH         NEW HAMPSHIRE  ...    33   4517     68
31  6/26/2021   NJ            NEW JERSEY  ...    34  25458    880
32  6/26/2021   NM            NEW MEXICO  ...    35   7557    268
33  6/26/2021   NY              NEW YORK  ...    36  65611   1338
34  6/26/2021   NC        NORTH CAROLINA  ...    37  42683    756
35  6/26/2021   ND          NORTH DAKOTA  ...    38   3184     36
36  6/26/2021   MP   NORTHERN MARIANA IS  ...    69     19      0
37  6/26/2021   OH                  OHIO  ...    39  38234    997
38  6/26/2021   OK              OKLAHOMA  ...    40   8605    596
39  6/26/2021   OR                OREGON  ...    41  23233    283
40  6/26/2021   PA          PENNSYLVANIA  ...    42  59790   1416
41  6/26/2021   PR           PUERTO RICO  ...    72   8472    243
42  6/26/2021   RI          RHODE ISLAND  ...    44   4328     57
43  6/26/2021   SC        SOUTH CAROLINA  ...    45  17485    322
44  6/26/2021   SD          SOUTH DAKOTA  ...    46   1832     63
45  6/26/2021   TN             TENNESSEE  ...    47  20933    349
46  6/26/2021   TX                 TEXAS  ...    48  98883   2046
47  6/26/2021   UT                  UTAH  ...    49  16351    153
48  6/26/2021   VT               VERMONT  ...    50   1440     10
49  6/26/2021   VI        VIRGIN ISLANDS  ...    78    710      3
50  6/26/2021   VA              VIRGINIA  ...    51  20090    622
51  6/26/2021   WA            WASHINGTON  ...    53  47355    401
52  6/26/2021   WV         WEST VIRGINIA  ...    54  10676    197
53  6/26/2021   WI             WISCONSIN  ...    55  16802    549
54  6/26/2021   WY               WYOMING  ...    56   3809     33

[55 rows x 10 columns]

##### Drop the redundant columns "date" and "state".

In [9]:
df5=df5.drop(["date","state"],axis=1)
df5.head()

Date code       State  ...  fips  cases  deaths
0  6/26/2021   AL     ALABAMA  ...     1  22012     440
1  6/26/2021   AK      ALASKA  ...     2   2887      27
2  6/26/2021   AZ     ARIZONA  ...     4  30214     590
3  6/26/2021   AR    ARKANSAS  ...     5  11529     154
4  6/26/2021   CA  CALIFORNIA  ...     6  72954    1799

[5 rows x 8 columns]

##### We are going to define the death rate as the percentage of COVID-19 patients who died. 

In [10]:
df5["death_rate(%)"]=df5["deaths"]*100/df5["cases"]
df5.head()

Date code       State  ...  cases  deaths  death_rate(%)
0  6/26/2021   AL     ALABAMA  ...  22012     440       1.998910
1  6/26/2021   AK      ALASKA  ...   2887      27       0.935227
2  6/26/2021   AZ     ARIZONA  ...  30214     590       1.952737
3  6/26/2021   AR    ARKANSAS  ...  11529     154       1.335762
4  6/26/2021   CA  CALIFORNIA  ...  72954    1799       2.465937

[5 rows x 9 columns]

In [11]:
df5=df5.sort_values(by=["Series_Complete_Pop_Pct"],ascending=False)

In [12]:
df5=df5.rename(columns={'Administered_Dose1_Pop_Pct':'% of population received 1st dose','Series_Complete_Pop_Pct':"% of population fully vaccinated"})

In [13]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

def vaccine_impact(col):
  fig = make_subplots(specs=[[{"secondary_y": True}]])
  fig.add_trace(
    go.Bar(x=df5['State'], y=df5[col], name="Percentage of popuplation received the first dose of COVID vaccine"),
    secondary_y=False,)
  fig.add_trace(
    go.Scatter(x=df5['State'], y=df5['death_rate(%)'], name='death_rate(%)'),
    secondary_y=True,)
  fig.update_layout(width=1600)
  fig.update_xaxes(title_text="State")
  fig.update_yaxes(title_text=f"<b>primary</b> {col} ", secondary_y=False)
  fig.update_yaxes(title_text="<b>secondary</b> death_rate(%)", secondary_y=True)
  return fig.show()

In [ ]:
print("Impact of first vaccine dose on death rate")
vaccine_impact('% of population received 1st dose')

Impact of first vaccine dose on death rate


In [14]:
print("Impact of full vaccination on death rate")
vaccine_impact('% of population fully vaccinated')

Impact of full vaccination on death rate


##### Death rate is the percentage of COVID patients that died after 5/1/2021. Vaccines had been made available to all people by this time. Did universal eligibility for vaccine help lower death rate? Death rate since 5/1/2021 is below 7% in all states. Every 100 COVID patients, at most 7 people died. However, the graphs show that high percentage of vaccinated population does not correlate to low percentage of deaths among COVID patients. There are some relatively low percentage of fully vaccinated people and low death rate since 5/1/2021, such as Wyoming. In the state of Wyoming, only 34.1% of the population is fully vaccinated, a low percentage compared to states such as Vermont, Massachusetts, Maine, Connecticut, Rhode Island. However, Wyoming also has one of the lowest COVID death rate in the country. Since 5/1/2021, just less than 1% of the COVID patient in the death died. In the state of Maryland, 54.6% of the population are fully vaccinated, a very high percentage. However, Maryland also has the second highest death rate since 5/2/2021. 6.7% of the COVID patients in Maryland died. States with high percentage of fully vaccinated population do not all have low death rates since 5/1/2021. That could be due to the fact that some states had high death rates to begin with; Tt could also be due to the different public health measures each state had.

In [ ]:
df2["death_rate(%)"]=df2["deaths"]*100/df2["cases"]

import plotly.express as px
import numpy as np

for s in list(np.unique(df2["state"])):
  df2_s=df2[df2["state"]==s]
  fig = px.scatter(df2_s,x="date", y="death_rate(%)",title=f"Percentage of COVID patients died in {s} since the onset of pandemic")
  fig.show()


##### Cumulatively for all states, death rate, the percentage of COVID patients who died, plateaued around 3/1/2021, a timeframe that vaccine became available to most people. 

##### Combining the analytical results from above (the bar charts and the scatter plots), we arrive at the following conclusions: (1) vaccine is the major tool in the fight against COVID-19; (2) the fact that after 5/1/2021, a timeframe when vaccine has been made available to all people, some states with high percentage of fully vaccinated population still have high death rates (as compared to other states), or the fact that in some states fully vaccinated people account for less than 40% of the population, tells us that it is not time to relax all the COVID health guidance yet. Guidances such as masks and social distancing are still necessary.